core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java

```diff
@@ -300,7 +300,8 @@ private void advanceToNextPage() {
     @Override
     public boolean hasNext() {
       if (numRecords == 0) {
-        if (reader != null) {
+        // hasNext maybe called multiple times. Let's guard that
+        if (reader != null && spillWriters.size() > 0) {
```

Comment:
This is probably OK but is this the right condition to check? and should handleFailedDelete() do this check?

Reply:
Yeah, I thought about that. It's weird to do the check here. The better solution would be add a new field to indicate the iterator has reached to its end, but seems quite overkill.

How do you like the following?
```
    private void handleFailedDelete() {
      if (spillWriters.size() > 0) {
        // remove the spill file from disk
        File file = spillWriters.removeFirst().getFile();
        if (file != null && file.exists() && !file.delete()) {
          logger.error("Was unable to delete spill file {}", file.getAbsolutePath());
        }
      }
    }

```

Reply:
Yeah that looks a little more direct and tidy, if it also solves the issue.